compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ForLoopsLowering.kt

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.common.lower.loops
+
+import org.jetbrains.kotlin.backend.common.CommonBackendContext
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.irIfThen
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.irCallOp
+import org.jetbrains.kotlin.ir.builders.irGet
+import org.jetbrains.kotlin.ir.builders.irSetVar
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrVariable
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrCompositeImpl
+import org.jetbrains.kotlin.ir.symbols.IrVariableSymbol
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.toKotlinType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+
+val forLoopsPhase = makeIrFilePhase(
+    ::ForLoopsLowering,
+    name = "ForLoopsLowering",
+    description = "For loops lowering"
+)
+
+/**
+ * This lowering pass optimizes for-loops.
+ *
+ * Replace iteration over progressions (e.g., X.indices, a..b) and arrays with
+ * a simple while loop over primitive induction variable.
+ *
+ * For example, this loop:
+ *
+ * ```
+ * for (i in first..last step foo) { ... }
+ * ```
+ *
+ * is represented in IR in such a manner:
+ *
+ * ```
+ * val it = (first..last step foo).iterator()
+ * while (it.hasNext()) {
+ *     val i = it.next()
+ *     ...
+ * }
+ * ```
+ *
+ * We transform it into the following loop:
+ *
+ * ```
+ * var it = first
+ * if (it <= last) {  // (it >= last if the progression is decreasing)
+ *     do {
+ *         val i = it++
+ *         ...
+ *     } while (i != last)
+ * }
+ * ```
+ *
+ * In case of iteration over array we transform it into following:
+ *
+ * ```
+ * while (it <= array.size - 1) {
```

Comment:
Please add bytecode tests

Reply:
Seems jvm backend should have different behaviour for such optimizaion on array reassignment inside loops depending of compiler version, @dnpetrov correct me if not. Maybe it shouldn't be supported in IR

Reply:
@max-kammerer Reassignment of array loop variables has changed semantics in 1.3 ([KT-21354](https://youtrack.jetbrains.com/issue/KT-21354)) and we don't intend (yet, at least) to support `-language-version 1.2` in the IR backend, so I think we should only support the new semantics.

Reply:
> Please add bytecode tests

@max-kammerer Done!

> Reassignment of array loop variables has changed semantics in 1.3 (KT-21354) ... we should only support the new semantics.

@udalov @max-kammerer **Please confirm:** The optimization in this PR _only_ supports the new semantics (since a temporary variable for the array is created). This is my understanding if I'm reading KT-21354 (and what you said) correctly.

Reply:
Yes, this is correct. Since the tests on the new semantics pass in the IR backend in master and are not changed in this optimization, the new semantics is supported.